### DISCUSSION TITLE
Get types for inline fragments

### DISCUSSION THREAD

**Discussion by ThomasKientz** [0no-co/gql.tada#392] [Q&A] [answered]
Hi, 

What am I missing to get this properly typed ? 

```typescript
export const createOnboarding = async () => {
  const query = graphql(`
    mutation MyQuery {
      onboardCompanyAccountHolder(
        input: { companyType: Company, isRegistered: true, accountName: "12 rue" }
      ) {
        ... on OnboardCompanyAccountHolderSuccessPayload {
          onboarding {
            id
          }
        }
      }
    }
  `);

  const { data } = await swanApi.mutation(query, {});

  return data?.onboardCompanyAccountHolder.onboarding.id; //  TS error : Property 'onboarding' does not exist
                                                                                 
};
```

<img width="1257" alt="Capture d’écran 2024-09-12 à 20 44 50" src="https://github.com/user-attachments/assets/dd94997f-f2ad-4f47-a2fd-4451571da83a">

**Comment by JoviDeCroock** [comment]
That info is a bit insufficient but I'll assume that the field you are spreading a fragment on is abstract. In that case you need to narrow the type by doing __typename === typeCondition

**Reply by ThomasKientz** [reply]
Sorry for the lack of context, but I am new to graphQl. I can run the query on the playground (screenshot), I don't know really know what is an abstract fragment. What information can I provide ? What do you mean by "doing __typename === typeCondition" ?

**Reply by ThomasKientz** [reply]
I don't really know whats the field I am spreading on is. Here is what the [documentation](https://docs.swan.io/developers/graphql#graphql-mutations-example) says 

> Consider the following example, which calls the mutation updateAccount to update the account's language. Note the ... on UpdateAccountSuccessPayload, in which you can choose to return the account's language in the payload. Swan recommends adding validations systematically, as well as rejections.

**Reply by JoviDeCroock** [reply, marked as answer]
The fact that you are using an inline-fragment `... on X` here is because `onboardCompanyAccountHolder` is returning an abstract type which is either the success or error variant. You are asking it to show you the success variant and for errors you won't receive anything.

Now TypeScript ofcourse knows that this can have two variants so you'll have to narrow the type by doing the following

```gql
    mutation MyQuery {
      onboardCompanyAccountHolder(
        input: { companyType: Company, isRegistered: true, accountName: "12 rue" }
      ) {
        __typename ## notice this, this can be OnboardCompanyAccountHolderSuccessPayload or the error variant
        ... on OnboardCompanyAccountHolderSuccessPayload {
          onboarding {
            id
          }
        }
      }
    }
```

NOw if you do this in TS

```ts
if (data.onboardCompanyAccountHolder.__typename === 'OnboardCompanyAccountHolderSuccessPayload')
```

The type will be narrowed, if you would hover the `onboardCompanyAccountHolder` it would have shown you this in the hover-information as well as it would show two objects one with `OnboardCompanyAccountHolderSuccessPayload` and one with the error variant.

When I do CMD + Enter on the TypeCondition of that fragment I see all possible types that can be returned for said abstract type:

<img width="244" alt="Screenshot 2024-09-12 at 22 16 23" src="https://github.com/user-attachments/assets/8465b247-a90b-403c-aac1-9c8b4a485d20">

**Reply by ThomasKientz** [reply]
Okay, got it. That makes total sense. Thanks a lot!
